dig/README.md

```diff
@@ -0,0 +1,78 @@
+# Dependency Injection Graph
+
+This package provides a fairly opinionated way of resolving object dependencies.
+There are two sides of dig: `Inject` and `Resolve`
+
+## Inject
+
+Adding an object, or a constructor of an object to the graph.
+
+There are two ways of providing an object:
+
+1. Injecting a pointer to an existing object
+1. Injecting a "constructor function" that returns one pointer (or interface)
+
+### Injecting an object
+
+Injecting an object means it has no dependencies, and will be used as a
+**shared** singleton instance for all resolutions within the graph
+
+```go
+type Fake struct {
+    Name string
+}
+err := g.Inject(&Fake{Name: "I am an injected thing"})
+require.NoError(t, err)
+
+var f1 *Fake
+err = g.Resolve(&f1)
+require.NoError(t, err)
+
+// f1 is ready to use here...
+```
+
+### Injecting a constructor
+
+This is a more interesting and widely used scenario. Constructor is defined as a
+function that returns exactly one pointer (or interface) and takes 0-N number of
+arguments. Each one of the arguments is automatically registered as a
+**dependency** and must also be an interface or a pointer.
+
+The following example illustrates injecting a constructor function for type
+`*Object` that requires `*Dep` to be present in the graph
+
+```go
+type Dep struct{}
+
+type Object struct{
+  Dep
+}
+
+func NewObject(d *Dep) *Object {
```

Comment:
can constructors return an error? it seems reasonable since resolve returns an error.